NKL/Encode.lean

```diff
@@ -151,117 +150,202 @@ private def decList (f : DecodeM a) : DecodeM (List a) := do
 -- Expressions
 
 partial def encExpr : Expr -> ByteArray
-  | .value c          => tag 0x10 [encConst c]
-  | .bvar s           => tag 0x11 [encString s]
-  | .var s _          => tag 0x12 [encString s]
-  | .subscript e ix   => tag 0x13 [encExpr e, encList encExpr ix]
-  | .slice l u step   => tag 0x14 [encExpr l, encExpr u, encExpr step]
-  | .binop op l r     => tag 0x15 [encString op, encExpr l, encExpr r]
-  | .cond c t e       => tag 0x16 [encExpr c, encExpr t, encExpr e]
-  | .tuple es         => tag 0x17 [encList encExpr es]
-  | .list es          => tag 0x18 [encList encExpr es]
-  | .call f ax        => tag 0x19 [encExpr f, encList encExpr ax]
-  | .gridcall f ix ax => tag 0x1a [encExpr f, encList encExpr ix, encList encExpr ax]
+  | .var s            => tag 0x30 [encString s]
+  | .tensor t s       => tag 0x31 [encString t, encList encInt s]
+  | .const c          => tag 0x32 [encConst c]
+  | .tuple es         => tag 0x33 [encList encExpr es]
+  | .list es          => tag 0x34 [encList encExpr es]
+  | .access e ix      => tag 0x35 [encExpr e, encList encIndex ix]
+  | .binop op l r     => tag 0x36 [encString op, encExpr l, encExpr r]
+  | .unop op e        => tag 0x37 [encString op, encExpr e]
+  | .call f ax kw     => tag 0x38 [encExpr f, encList encExpr ax, encList encKeyword kw]
+where
+  encKeyword : String × Expr -> ByteArray
+  | (key, expr) => (encString key).append (encExpr expr)
 
 partial def decExpr : DecodeM Expr := do
   match (<- next) with
-  | 0x10 => return .value (<- decConst)
-  | 0x11 => return .bvar (<- decString)
-  | 0x12 => return .var (<- decString) ""
-  | 0x13 => return .subscript (<- decExpr) (<- decList decExpr)
-  | 0x14 => return .slice (<- decExpr) (<- decExpr) (<- decExpr)
-  | 0x15 => return .binop (<- decString) (<- decExpr) (<- decExpr)
-  | 0x16 => return .cond (<- decExpr) (<- decExpr) (<- decExpr)
-  | 0x17 => return .tuple (<- decList decExpr)
-  | 0x18 => return .list (<- decList decExpr)
-  | 0x19 => return .call (<- decExpr) (<- decList decExpr)
-  | 0x1a => return .gridcall (<- decExpr) (<- decList decExpr) (<- decList decExpr)
+  | 0x30 => return .var (<- decString)
+  | 0x31 => return .tensor (<- decString) (<- decList decInt)
+  | 0x32 => return .const (<- decConst)
+  | 0x33 => return .tuple (<- decList decExpr)
+  | 0x34 => return .list (<- decList decExpr)
+  | 0x35 => return .access (<- decExpr) (<- decList decIndex)
+  | 0x36 => return .binop (<- decString) (<- decExpr) (<- decExpr)
+  | 0x37 => return .unop (<- decString) (<- decExpr)
+  | 0x38 => return .call (<- decExpr) (<- decList decExpr) (<- decList decKeyword)
   | t => throw s!"Unknown tag in Expr {t}"
+where
+  decKeyword : DecodeM (String × Expr) :=
+    return ((<- decString), (<- decExpr))
 
 private def chkExpr (e : Expr) : Bool :=
   (decode' decExpr $ encExpr e) == some e
 
-private def nil := Expr.value .nil
+private def nil := Expr.const .none
```

Comment:
Call this `none`?

Reply:
I was trying to avoid clashing with Lean's none.